src/main/scala/devices/spi/SPIPhysical.scala

```diff
@@ -59,11 +112,22 @@ class SPIPhysical(c: SPIParamsBase) extends Module {
     Mux(fmt.endian === SPIEndian.MSB, data, Cat(data.toBools))
 
   val rxd = Cat(io.port.dq.reverse.map(_.i))
-  val samples = Seq(rxd(1), rxd(1, 0), rxd)
+  val rxd_delayed = Vec(Seq.fill(io.port.dq.size)(false.B))
+
+  //Adding fine-granularity delay buffers on the received data
+  val fine_grain_delay = Seq.fill(io.port.dq.size) {Module(new BlackBoxDelayBuffer())}
+  for (j <- 0 to 3 ) { 
```

Comment:
since fine_grain_delay may have more or less than 4 elements, can you use the same `io.port.dq.size` parameter there?

You could also use a `foreach`

Reply:
okay, changed the constant to parameter. I am not very familiar with using foreach though. Is there any distinct advantage over using 'for'?